### PR TITLE
Plugin E2E: Enable testing the query editor in Explore

### PR DESCRIPTION
**Pull request by sunker**
**What this PR does / why we need it**:
This PR adds fixture and model that enable testing a data source query editor and query runner in Explore. Also adding a few tests that showcases how Explore tests can be written using this feature.

**Which issue(s) this PR fixes**:

Part of https://github.com/grafana/plugin-tools/issues/576

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: packages/plugin-e2e/src/models/GrafanaPage.ts [hunk 1]
```diff
@@ -9,6 +9,16 @@ import { PluginTestCtx } from '../types';
 export abstract class GrafanaPage {
   constructor(public readonly ctx: PluginTestCtx, protected readonly expect: Expect<any>) {}
 
+  protected async navigate(url: string, options?: NavigateOptions) {
```
Comment (sunker) [inline packages/plugin-e2e/src/models/GrafanaPage.ts:12]: Planning to use this method across all models, but will do that in a separate PR.

File: packages/plugin-e2e/src/models/TimeRange.ts [hunk 2]
```diff
@@ -27,18 +27,5 @@ export class TimeRange extends GrafanaPage {
     await toField.clear();
     await toField.fill(to);
     await this.getByTestIdOrAriaLabel(this.ctx.selectors.components.TimePicker.applyTimeRange).click();
-
-    await this.expect
```
Comment (sunker) [inline packages/plugin-e2e/src/models/TimeRange.ts:31]: Don't think we need these soft expects.

File: packages/plugin-e2e/src/models/ExplorePage.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    });
+    await this.expect(locator).toBeVisible();
+    return locator;
+  }
+
+  async runQuery(options?: RequestOptions) {
+    const components = this.ctx.selectors.components;
+    const runButtonTextContent = await this.getByTestIdOrAriaLabel(components.RefreshPicker.runButtonV2).textContent();
+    // if there is a query running, wait for it to finish before starting a new one
+    if (runButtonTextContent === 'Cancel') {
```
Comment (academo) [inline packages/plugin-e2e/src/models/ExplorePage.ts:33]: is there a way with playwright to check if there's a request to this url ongoing? or use the waitForResponse to return if no request is going?

File: packages/plugin-e2e/src/models/ExplorePage.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    const responsePromise = this.ctx.page.waitForResponse((resp) => resp.url().includes('/api/ds/query'), options);
+    try {
+      await this.getByTestIdOrAriaLabel(components.RefreshPicker.runButtonV2).click({
+        timeout: 1000,
+      });
+    } catch (_) {
+      // handle the case when the run button is hidden behind the "Show more items" button
+      await this.getByTestIdOrAriaLabel(components.PageToolbar.item('Show more items')).click();
```
Comment (academo) [inline packages/plugin-e2e/src/models/ExplorePage.ts:44]: not sure how you are handling these magic strings so far, but would it be better to put all these in a file where we can re-use them and have more information related to it.

e.g. `PageToolbar.ShowMoreItemsButton`

This is because the string alone doesn't give context of where is this string supposed to be in the UI and less about being able to re-use or easily update.

File: packages/plugin-e2e/src/models/ExplorePage.ts (hunk 3 above)
Comment (sunker) [reply packages/plugin-e2e/src/models/ExplorePage.ts:33]: Yeah I agree this is not ideal. I'll look into ways of improving this tomorrow.

File: packages/plugin-e2e/src/models/ExplorePage.ts (hunk 4 above)
Comment (sunker) [reply packages/plugin-e2e/src/models/ExplorePage.ts:44]: Yeah this is a huge anti-pattern that I've come across a few times in the grafana/e2e-selectors package. The whole point of this package is to provide a safe way to access these selectors, but with these functions with magic strings we're kind of losing the whole purpose of it. 

I think rewriting all of these selectors in grafana/e2e-selectors will be a too big undertaking for now, but I'll make note in the tasklist to at least change the ones that we use in grafana-e2e (I know they're being used at least two more times).

Comment (academo) [reply packages/plugin-e2e/src/models/ExplorePage.ts:44]: I am not suggesting we move them to e2e/selectors or a whole different package. something as simple as defining the variable at the beginning of the file to avoid this to turn into a  [magic string](https://en.wikipedia.org/wiki/Magic_string)

File: packages/plugin-e2e/src/models/ExplorePage.ts (hunk 3 above)
Comment (sunker) [reply packages/plugin-e2e/src/models/ExplorePage.ts]: Networking is an area where Playwright falls a bit short IMO. To achieve something like this, we'd need to setup one event listener that keeps track of new requests, and a second one that keeps track of responses. Then you'd need your own mechanism that correlates these two. 

It seems like the sheets data source is triggering a `onRunQuery` after any change in the query editor. For that reason, there may already be an ongoing request when run query method is executed in the test. This is rather unconventional, so after closer consideration I don't think this should be handled in the lib, but in the data source instead. I'm removing this code.

File: packages/plugin-e2e/src/models/ExplorePage.ts (hunk 4 above)
Comment (sunker) [reply packages/plugin-e2e/src/models/ExplorePage.ts]: Done

File: packages/plugin-e2e/src/models/ExplorePage.ts (hunk 3 above)
Comment (academo) [reply packages/plugin-e2e/src/models/ExplorePage.ts]: sounds good! the current approach will have to work I guess :)

**Review by academo — APPROVED** [review]
